# Exercise 2
Add the specified code for each code cell, running the cells _in order_.

Open the included **`data/words_small.txt`** file (containing a "small" list of English words) and `read()` it into a single string variable **`words_str`**. Then print the first 100 _characters_ of the `words_str` variable (use a slice operator!)

In [8]:
with open('data/words_small.txt') as file: 
    words_str = file.read()
print(words_str[:100])

abbreviate
abbreviation
abdomen
able
aboard
abroad
absent
absolute
abuse
academic
academy
accelerate


**Using a loop**, count the number of times the newline character `\n` appears in the `words_str`. This should tell you the number of words in the file (plus or minus one). You can check your work with the built-in `count()` string method (but use the loop for practice!)

In [20]:
count = 0
for char in words_str: 
    if char == '\n': 
        count += 1
print(count)

#test using count()
print(words_str.count('\n'))

1520
1520


Create a variable **`letter_counts`** that is a dictionary whose keys are characters (letters) and whose values are the number of times that character appears in `words_str`. Print the dictionary after you make it.
- To do this, start with an empty dictionary, then loop through the letters one at a time. At each letter, increase the value associated with that letter key by one.
- Watch out for if the letter is not in the dictionary yet!
- Also _delete_ the entry for the newline `\n` character

In [47]:
letter_counts = {}
for letter in words_str: 
    ###get previous count
    previous_count = letter_counts.get(letter, 0)
    ###incrememtn prev cnt by 1
    letter_counts[letter] = previous_count + 1

##remove '\n'    
letter_counts.pop('\n')   
print(letter_counts)


{'a': 703, 'b': 188, 'r': 642, 'e': 1118, 'v': 85, 'i': 515, 't': 604, 'o': 598, 'n': 550, 'd': 314, 'm': 223, 'l': 457, 's': 504, 'u': 279, 'c': 328, 'y': 189, 'p': 224, 'h': 295, 'f': 155, 'w': 161, 'g': 213, 'k': 126, 'z': 15, 'x': 23, 'j': 16, 'q': 9}


Print out all the count in the `letter_counts` dictionary in the format `"A appears 703 times"`, each on a new line, in alphabetical order. _Hint_: iterate through the sorted keys!

In [50]:
##Alternative approach
# sorted_keys = sorted(letter_counts.keys())
# for key in sorted_keys: 
#     print(key,"appears in", letter_counts[key], "times")
  
for key, value in sorted(letter_counts.items()): 
    print(key,"appears in", value, "times")

a appears in 703 times
b appears in 188 times
c appears in 328 times
d appears in 314 times
e appears in 1118 times
f appears in 155 times
g appears in 213 times
h appears in 295 times
i appears in 515 times
j appears in 16 times
k appears in 126 times
l appears in 457 times
m appears in 223 times
n appears in 550 times
o appears in 598 times
p appears in 224 times
q appears in 9 times
r appears in 642 times
s appears in 504 times
t appears in 604 times
u appears in 279 times
v appears in 85 times
w appears in 161 times
x appears in 23 times
y appears in 189 times
z appears in 15 times


Print which letter from the `letter_counts` dictionary has a **value** of `515`. _Hint:_ use a loop to look through all the _items_ until you find the right one, then print that!

In [52]:
for letter, count in letter_counts.items(): 
    if count == 515: 
        print(letter)

i


Create a variable **`count_to_letter_dict`** that is a dictionary whose **keys** are the counts and whose **values** are the letters that have that count. This is a _reverse lookup_ dictionary. Hint: ceate a new dictionary, then loop through items in `letter_counts` and add an item to your new dictionary for each element. You can assume no duplicated counts.

In [89]:
count_to_letter_dict = {}
for letter, count in letter_counts.items(): 
    count_to_letter_dict[count] = letter
count_to_letter_dict
#count_to_letter_dict[642]

{703: 'a',
 188: 'b',
 642: 'r',
 1118: 'e',
 85: 'v',
 515: 'i',
 604: 't',
 598: 'o',
 550: 'n',
 314: 'd',
 223: 'm',
 457: 'l',
 504: 's',
 279: 'u',
 328: 'c',
 189: 'y',
 224: 'p',
 295: 'h',
 155: 'f',
 161: 'w',
 213: 'g',
 126: 'k',
 15: 'z',
 23: 'x',
 16: 'j',
 9: 'q'}

Print the **10** most common letters and their counts with the format `"A appears 703 times"`. _Hint:_ you'll need to sort the keys to `count_to_letter_dict` and take the first 10 of those.

In [95]:
sorted_keys = sorted(count_to_letter_dict.keys(), reverse = True)
for count in sorted_keys[:10]: 
    print(count_to_letter_dict[count].upper(), "appears", count, "times")

[1118, 703, 642, 604, 598, 550, 515, 504, 457, 328]
E appears 1118 times
A appears 703 times
R appears 642 times
T appears 604 times
O appears 598 times
N appears 550 times
I appears 515 times
S appears 504 times
L appears 457 times
C appears 328 times


Bonus: repeat your analysis on the `data/words_huge.txt` file! Are the top letters the same?